From 5fb22c3b5cd8c5b5eb38bd215b005ad9f2161831 Mon Sep 17 00:00:00 2001
From: Alexander Barabash <alexander_barabash@mentor.com>
Date: Tue, 8 Dec 2015 15:09:23 +0200
Subject: [PATCH 05/11] Support for gst_flow_event LLTng tracepoints.

The gst_flow_event tracepoints trace the properties of events sent through the gstreamer pipeline.
This replaces gst_flow_data tracepoints for event data.

The initial support for this kind of tracepoints includes calling appropriate primitives
when the data is passed through the pipeline elements, in the following APIs:

   gst_pad_push_event()
   gst_pad_send_event()

The tracepoints are invoked just before the gst_flow tracepoints.

At this stage, only minimal data is stored for each tracepoint:

    -- a unique integer ID of the invoking thread;
    -- the value of the pointer to the event.
    -- the type of event.
---
 gst/gst_tracepoints.h | 11 +++++++++++
 gst/gstpad.c          |  1 +
 2 files changed, 12 insertions(+)

diff --git a/gst/gst_tracepoints.h b/gst/gst_tracepoints.h
index 5267af4..15d85f5 100644
--- a/gst/gst_tracepoints.h
+++ b/gst/gst_tracepoints.h
@@ -59,6 +59,7 @@ typedef enum GstFlowTracepointDataType {
 #define ctf_gst_tracepoint_type_field(name, type) ctf_integer (guint8, name, (guint8) (type))
 #define ctf_gst_tracepoint_kind_field(name, kind) ctf_integer (guint8, name, (guint8) (kind))
 #define ctf_gst_data_type_field(name, type) ctf_integer (guint8, name, (guint8) (type))
+#define ctf_gst_event_type_field(name, event) ctf_integer (gint, name, (gint) (event->type))
 
 #endif /* _GST_TRACEPOINTS_H */
 
@@ -96,6 +97,12 @@ GST_TRACEPOINT_EVENT (gst_flow_data,
                                  ctf_gst_thread_id_field (thread_id)
                                  ctf_gst_data_type_field (data_type, ((trace_is_on ? (*trace_is_on = 1) : 0), data_type))))
 
+GST_TRACEPOINT_EVENT (gst_flow_event,
+                      TP_ARGS (GstEvent *, event),
+                      TP_FIELDS (ctf_gst_data_field (event, event)
+                                 ctf_gst_thread_id_field (thread_id)
+                                 ctf_gst_event_type_field (event_type, event)))
+
 #endif /* _GST_TRACEPOINTS_H */
 
 #include <lttng/tracepoint-event.h>
@@ -139,6 +146,9 @@ GST_TRACEPOINT_EVENT (gst_flow_data,
 #define GST_FLOW_TRACEPOINT_BUFFER(buffer) \
   GST_TRACEPOINT (gst_flow_data, buffer, GST_TRACEPOINT_DATA_TYPE_BUFFER, NULL)
 
+#define GST_FLOW_TRACEPOINT_EVENT(event) \
+  GST_TRACEPOINT (gst_flow_event, event)
+
 #define GST_FLOW_TRACEPOINT_DATA(data, is_buffer)                       \
     do {                                                                \
       int trace_is_on = 0;                                              \
@@ -174,6 +184,7 @@ GST_TRACEPOINT_EVENT (gst_flow_data,
 #define GST_FLOW_TRACEPOINT_SEND_EVENT_ERROR(pad)
 #define GST_FLOW_TRACEPOINT_SEND_EVENT_DROPPED(pad)
 #define GST_FLOW_TRACEPOINT_BUFFER(buffer)
+#define GST_FLOW_TRACEPOINT_EVENT(event)
 #define GST_FLOW_TRACEPOINT_DATA(data, is_buffer)
 #define GST_FLOW_TRACEPOINT_BUFFER_LIST(list)
 
diff --git a/gst/gstpad.c b/gst/gstpad.c
index 8916da5..579fb1c 100644
--- a/gst/gstpad.c
+++ b/gst/gstpad.c
@@ -5388,6 +5388,7 @@ gst_pad_send_event_unchecked (GstPad * pad, GstEvent * event,
   GstObject *parent;
 
   GST_OBJECT_LOCK (pad);
+  GST_FLOW_TRACEPOINT_EVENT (event);
   GST_FLOW_TRACEPOINT_SEND_EVENT_ENTER (pad, event);
 
   event = apply_pad_offset (pad, event, GST_PAD_IS_SRC (pad));
-- 
2.2.1

